flink/src/main/java/org/apache/iceberg/flink/source/FlinkInputFormat.java

```diff
@@ -72,12 +62,7 @@ public BaseStatistics getStatistics(BaseStatistics cachedStatistics) {
 
   @Override
   public FlinkInputSplit[] createInputSplits(int minNumSplits) throws IOException {
-    // Called in Job manager, so it is OK to load table from catalog.
-    tableLoader.open();
```

Comment:
@stevenzwu @openinx, this is something we talked a little bit about in #2978.

After a closer look, this would actually change the behavior. As of now, the table is loaded two times: in `FlinkSource` and in `FlinkInputFormat`. After this change, we only load it in `FlinkSource` and then pass a read-only serializable copy to `FlinkInputFormat`.

What is the lifecycle of `FlinkInputFormat`? Is it alright to pass a read-only table to it that won't pick up any new snapshots or table properties?

Reply:
As I stated above, I've put Flink and MR related changes into this PR just temporarily.

Instead of passing `Table`, I can add one more parameter and pass a partition type alongside other existing parameters. Using `Table` seems cleaner but not a blocker for my current work.

Reply:
I had to modify a test in `TestFlinkScanSql`.

Reply:
not an expert in Flink, but agree this is a change of behavior that we should be cautious. Currently it seems that the input splits are created with the freshly loaded table from the serializable table loader. This means that a table can potentially be updated during the process and then generate a different array of splits comparing to using a read-only copy. 

I am not sure which behavior is correct though, it feels to me that using a read-only copy is more valid to ensure snapshot isolation. But if the Flink source is supposed to be reused to reflect always the latest view of the data, then using the table loader becomes necessary.

Reply:
> What is the lifecycle of FlinkInputFormat? Is it alright to pass a read-only table to it that won't pick up any new snapshots or table properties?

@aokolnychyi  I think it is ok to pass a read-only table to FlinkInputFormat. it is used in two scenarios
1) batch mode: split discovery is done once at startup
2) streaming mode: it is used only in reader. dynamic/periodical split discovery is done by `StreamingMonitorFunction`, which would require a TableLoader to refresh table state

Reply:
Okay, I've submitted #2987 for Flink changes alone.

Reply:
Resolving, no longer applies.